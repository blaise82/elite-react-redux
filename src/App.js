import React, { Component } from 'react';
import { connect } from "react-redux";
import { loginAction } from './redux/actions/loginAction';
class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      emailError: '',
      passwordError: '',
      emailErrorStatus: false,
      passwordErrorStatus: false,
    };
  }

  validate = () => {
    let isError = false;
    const errors = {
      emailError: '',
      passwordError: '',
      emailErrorStatus: false,
      passwordErrorStatus: false,
    };

    if (this.state.email.indexOf('@') === -1) {
      isError = true;
      errors.emailErrorStatus = true;
      errors.emailError = 'Provide a valid email';
    }
    if (this.state.password.length == 0) {
      isError = true;
      errors.passwordErrorStatus = true;
      errors.passwordError = 'Provide a password';
    }
    this.setState({
      ...this.state,
      ...errors,
    });
    return isError;
  };

  onSubmit = async (e) => {
    e.preventDefault();
    const error = this.validate();
    if (!error) {
      const data = {
        email: this.state.email,
        password: this.state.password
      }
      await this.props.loginAction(data)
    } 
  };

  change = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    return (
      <div>
        <p style={{color: "green", display: this.props.loginState.loading}} className="loading">loading.....</p>
        <form onSubmit={(e) => this.onSubmit(e)} className="form">
          <h1>Sign in</h1> <br />
          <span style={{color: "red"}}>{ this.props.loginState.error}</span> <br/>
          <label htmlFor="">Email</label>
          <br />
          <input
            type="email"
            name="email"
            aria-label="email"
            placeholder="email"
            onChange={(e) => this.change(e)}
          />
          <br/>
            <span style={{color: "red"}}>{ this.state.emailError}</span>
          <br /> <br />
          <label htmlFor="">Password</label> <br />
          <input
            type="password"
            name="password"
            aria-label="password"
            placeholder="password"
            onChange={(e) => this.change(e)}
          />
          <br/>
          <span style={{color: "red"}}>{ this.state.passwordError}</span>
          <br />
          <input type="submit" value="LOGIN" aria-label="login" />
        </form>
      </div>
    );
  }
}


const mapStateToProps = (state) => {
  return {
    loginState: state.login
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    loginAction: (data) => dispatch(loginAction(data))
  }
}
export default connect(mapStateToProps,mapDispatchToProps)(App);
