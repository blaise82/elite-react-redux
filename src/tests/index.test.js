import React from 'react';
import {
  render,
  cleanup,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import App from '../App';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import store from '../redux/store';
const onSubmit = jest.fn();

const AppComp = () => {
  return render(
    <Provider store={store}>
      <App onSubmit={onSubmit} />
    </Provider>
  );
};
describe('APP', () => {
  afterEach(cleanup);

  it('should render the component', () => {
    const { asFragment } = AppComp();
    expect(asFragment(<App />)).toMatchSnapshot();
  });

  it('should change state/value of email on type', () => {
    const { getByLabelText } = AppComp();
    const email = getByLabelText('email');
    const password = getByLabelText('password');

    userEvent.type(email, 'email@gmail.com');
    userEvent.type(password, 'admin');

    expect(email.value).toBe('email@gmail.com');
    expect(password.value).toBe('admin');
  });

  it('should return error if password not provided', async () => {
    const { getByLabelText, container, getByText, debug } = AppComp();
    const email = getByLabelText('email');
    const form = container.querySelector('form');

    userEvent.type(email, 'email@gmail.com');

    form.dispatchEvent(new Event('submit'));

    waitFor(() => expect(getByText('Provide a password')).toBeTruthy());
  });

  it('should return error if password not provided', async () => {
    const { getByLabelText, container, getByText } = AppComp();
    const password = getByLabelText('password');
    const form = container.querySelector('form');

    userEvent.type(password, 'email@gmail.com');

    form.dispatchEvent(new Event('submit'));

    waitFor(() => expect(getByText('Provide a valid email')).toBeTruthy());
  });

  it('should return error if password and email are not correct', async () => {
    const { getByLabelText, container, getByText, debug } = AppComp();
    const password = getByLabelText('password');
    const email = getByLabelText('email');
    const button = getByLabelText('login');

    userEvent.type(email, 'email@gmail.com');
    userEvent.type(password, 'password');

    button.click(button);

    waitFor(() =>
      expect(getByText('Invalid Username or passoword')).toBeTruthy()
    );
  });
  //   it('should return backend error if password and email are not correct', () => {});

  //   it('should return backend error if password and email are not correct', () => {});
});
