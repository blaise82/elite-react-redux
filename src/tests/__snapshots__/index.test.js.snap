// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APP should render the component 1`] = `
<DocumentFragment>
  <div>
    <p
      class="loading"
      style="color: green; display: none;"
    >
      loading.....
    </p>
    <form
      class="form"
    >
      <h1>
        Sign in
      </h1>
       
      <br />
      <span
        style="color: red;"
      />
       
      <br />
      <label
        for=""
      >
        Email
      </label>
      <br />
      <input
        aria-label="email"
        name="email"
        placeholder="email"
        type="email"
      />
      <br />
      <span
        style="color: red;"
      />
      <br />
       
      <br />
      <label
        for=""
      >
        Password
      </label>
       
      <br />
      <input
        aria-label="password"
        name="password"
        placeholder="password"
        type="password"
      />
      <br />
      <span
        style="color: red;"
      />
      <br />
      <input
        aria-label="login"
        type="submit"
        value="LOGIN"
      />
    </form>
  </div>
</DocumentFragment>
`;
